nvtabular/torch_dataloader.py

```diff
@@ -218,20 +222,75 @@ def __len__(self):
         return self.num_chunks
 
     def __iter__(self):
-        buff = queue.Queue(1)
-        threading.Thread(target=self.load_chunk, args=(buff,)).start()
-        for _ in range(self.num_chunks):
-            chunk = buff.get()
-            yield from TensorItr(
-                chunk, batch_size=self.batch_size,
-            )
-
-    def load_chunk(self, out):
         for chunk in self.itr:
-            chunk = create_tensors_plain(
-                chunk, cat_cols=self.cat_cols, cont_cols=self.cont_cols, label_cols=self.label_cols
-            )
-            out.put(chunk)
+            yield chunk
+
+
+class ChunkQueue:
+    def __init__(
+        self,
+        num_chunks=2,
+        batch_size=1,
+        shuffle=False,
+        cat_cols=None,
+        cont_cols=None,
+        label_cols=None,
+    ):
+        self.num_chunks = num_chunks
+        self.batch_size = batch_size
+        self.q_in = queue.Queue(num_chunks)
+        self.q_out = queue.Queue(1)
+        self.cat_cols = cat_cols
+        self.cont_cols = cont_cols
+        self.label_cols = label_cols
+        self.shuffle = shuffle
+
+    def get(self):
+        return self.q_out.get()
+
+    def put(self, obj):
+        # check first for "end"
+        if isinstance(obj, str):
+            # clear the buffer
+            self.create_chunk(final=True)
+            # send bug out
+            self.q_out.put(obj)
+            return
+        self.q_in.put(obj)
+        if self.q_in.full():
+            self.create_chunk()
+
+    def create_chunk(self, final=False):
+        chunks = []
+        for _ in range(self.q_in.qsize()):
+            chunks.append(self.q_in.get())
+        if not chunks:
+            return
+        chunks = cudf.core.reshape.concat(chunks)
+        chunks.reset_index(drop=True, inplace=True)
+        if not final:
+            # stitching
+            chunks = self.stitch_run(chunks)
+            if not chunks or chunks.empty:
+                return
+        if self.shuffle:
+            _shuffle_gdf(chunks)
+        chunks = create_tensors_plain(
+            chunks, cat_cols=self.cat_cols, cont_cols=self.cont_cols, label_cols=self.label_cols
+        )
+        # chunk tensorized
+        self.q_out.put(chunks)
+
+    def stitch_run(self, chunks):
```

Comment:
I'm lost connecting `stitch_run` with what this function is doing (splitting a list of chunks up into one large chunk that's divisible by batch size and putting the "spill" back into the queue).

A better name might be something like `get_batch_size_divisible_chunk`?